egg/zoo/compo_vs_generalization/data.py

```diff
@@ -120,6 +124,65 @@ def split_train_test(dataset, p_hold_out=0.1, random_seed=7):
     return train, test
 
 
+def build_random_dataset(n_attributes: int,
+                         n_values: int,
+                         size: int,
+                         data_to_exclude: set=set(),
+                         allow_duplicates: bool=False,
+                         rng=None
+                         ) -> list:
+    """
+    Samples {size} elements from the input space, 
+    with no overlap with the set {data_to_exclude},
+    """
+    if n_values**n_attributes / size < 10e4 and allow_duplicates:
+        print(f"Warning : Building a dataset that probably contains duplicates")
+
+    # use a list if duplicates are allowed (uses less memory than set)
+    data = list() if allow_duplicates else set()
+    
+    while len(data) < size:
+        sample = tuple(torch.randint(n_values-1, size=(n_attributes,), generator=rng).tolist())
+        if sample not in data_to_exclude:
+            if allow_duplicates:
+                data.append(sample)
+            else:
+                data.add(sample)
+
+    return list(data)
+
+def build_datasets(n_attributes: int,
+                   n_values: int,
+                   train_size: int,
+                   test_size: int,
+                   validation_size: int,
+                   rng=None) -> Tuple[List, List, List]:
+    """
+    Returns the train, test and validation sets by sampling the data.
+    Samples can be repeated in the training set but not in the testing set
+    Also there is no overlap between train and test set.
+    """
+    # create test and validation sets simultaneously to prevent repeats
+    all_test_data = build_random_dataset(n_attributes,
+            n_values, 
+            size=test_size+validation_size,
+            allow_duplicates=False,
+            rng=rng)
+    test_set = all_test_data[:test_size]
+    validation_set = all_test_data[test_size:]
+
+    train_set = build_random_dataset(n_attributes,
+            n_values,
+            size=train_size,
+            data_to_exclude=set(test_set),
```

Comment:
This doesn't check for overlap with the validation set (which we also want to avoid, otherwise we will over-estimate our validation accuracy). You can join the two sets with `set(test_set) + set(validation_set)` I think (or use `all_test_data`)